### PR TITLE
Indexing changes for TMA Circular Buffering

### PR DESCRIPTION
**Pull request by rdspring1**
## Summary ##
It is the changes to the indexing lowering pass from https://github.com/NVIDIA/Fuser/pull/2773. It is stacked on #2824. 
Tracking Branch: https://github.com/NVIDIA/Fuser/pull/2773

## Details ##
- In the circular buffering pass, we manually index the mbarriers and tokens using the index of the prologue, main, and epilogue loops.
```cpp
for (int index : c10::irange(fl->extent()) {
  int stage = index % number_of_pipeline_stages;
  mbarrier_t current_stage_mbarrier = mbarriers[stage];  // represented with kir::TensorIndex

  int next_stage = (index + number_of_stages - 1) % number_of_pipeline_stages;
  mbarrier_t next_stage_mbarrier = mbarriers[next_stage];  // represented with kir::TensorIndex
}
```
- The handle functions for `kir::MBarrierInit`, `kir::MBarrierInvalidate`, `kir::MBarrierArriveExpectTx`, and `kir::MBarrierWait` are modified to handle `kir::TensorIndex`.
- `u32IndexScalarSmemTv` is modified to get the shared memory pointer address for a `kir::TensorIndex`.

### CODE REVIEW

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

File: csrc/device_lower/utils.h [hunk 1]
```diff
@@ -304,6 +304,10 @@ bool isExtentEqualToMaxParallelTypeExtent(const IterDomain* id);
 //! indexing special items in shared memory, like mbarrier.
 NVF_API Val* u32IndexScalarSmemTv(TensorView* tv);
 
+//! Get the uint32_t index of a TensorIndex. This is usually used for
```
Comment (naoyam) [inline csrc/device_lower/utils.h:307]: nit: I guess you copy-and-pasted the function above, but why is it called "scalar smem TV"? It seems we already has an address as a `Val`, so it doesn't seem to matter if it's scalar or not.

**Review by naoyam — COMMENTED** [review]
I'm a little confused. Previously there's no `handle` function for `MBarrierArriveExpectTx` and `MBarrierWait`. How were we lowering these ops?

**Comment by rdspring1** [comment]
Without circular buffering, `void IndexLowering::handleCpAsyncBulkLoad(const LoadStoreOp* ldst)` adds `MBarrierArriveExpectTx` and `MBarrierWait`. See https://github.com/NVIDIA/Fuser/blob/main/csrc/device_lower/pass/index.cpp#L1421-L1446

Everything is self-contained within the if-statement.
```cpp
  if (TDX == 0 && TDY == 0 && TDZ == 0) {
    __shared__ int64_t tokens;
    __shared__ int64_t mbarrier;
    init(mbarrier);
    MBarrierArriveExpectTx(mbarrier, data_size, num_threads);
    cp.async.bulk(data, mbarrier);
    MBarrierWait(mbarrier)
    inval(mbarrier);
 }
```

**Comment by naoyam** [comment]
Where is, for example, `kir::MBarrierArriveExpectTx` first inserted?

**Comment by rdspring1** [comment]
I'm not sure if you're referring to the circular buffering case, but `kir::MBarrierArriveExpectTx` is inserted at https://github.com/NVIDIA/Fuser/blob/main/csrc/device_lower/pass/index.cpp#L1434-L1435 for regular `cpAsyncBulk` load.

**Comment by naoyam** [comment]
> I'm not sure if you're referring to the circular buffering case, but `kir::MBarrierArriveExpectTx` is inserted at https://github.com/NVIDIA/Fuser/blob/main/csrc/device_lower/pass/index.cpp#L1434-L1435 for regular `cpAsyncBulk` load.

There should be a pass before IndexLowering that inserts `kir::MBarrierArriveExpectTx`, right? Otherwise, I don't understand why we would need `IndexLowering::hanlde(kir::MBarrierArriveExpectTx*)`. That handler should never be called if there's already `MBarrierArriveExpectTx` IR nodes.

**Comment by rdspring1** [comment]
The `kir::MBarrierArriveExpectTx` nodes are inserted in the circular buffering pass. Those changes are in the last PR #2833.
See https://github.com/NVIDIA/Fuser/pull/2833/files#diff-804c3cd07d9989909d30d986a6e04a00157713200cb74b255990d88cdc15ce44R663.

**Comment by naoyam** [comment]
> The `kir::MBarrierArriveExpectTx` nodes are inserted in the circular buffering pass. Those changes are in the last PR #2833. See https://github.com/NVIDIA/Fuser/pull/2833/files#diff-804c3cd07d9989909d30d986a6e04a00157713200cb74b255990d88cdc15ce44R663.

Ah, I see. Got it.

File: csrc/device_lower/pass/index.cpp [hunk 2]
```diff
@@ -1404,45 +1404,121 @@ void IndexLowering::handleGroupedGridWelford(
...
+    Expr* new_ldst =
+        IrBuilder::create<LoadStoreOp>(ldst->opType(), out, in, ldst->cacheOp())
+            ->withPredicate(ldst->predicate());
+    pushBack(new_ldst);
+
+    // register new LoadStoreOp with mbarrier
+    GpuLower::current()->ldstMBarrierIndexMap()[new_ldst] = mbarrier;
+
+    GpuLower::current()->propagateExprInfo(ldst, back());
+  } else {
```
Comment (jacobhinkle) [inline csrc/device_lower/pass/index.cpp:1495]: The circular buffering path will be tested in the tests included in #2773, but do we have an example that exercises this `else` branch?

File: csrc/device_lower/utils.h [hunk 3]
```diff
@@ -304,6 +304,10 @@ bool isExtentEqualToMaxParallelTypeExtent(const IterDomain* id);
...
+//! initializing a pipeline of mbarriers.
+NVF_API Val* u32IndexScalarSmemTv(kir::TensorIndex* index);
```
Comment (jacobhinkle) [inline csrc/device_lower/utils.h:309]: Why is this `NVF_API`?

File: csrc/device_lower/pass/index.cpp [hunk 4]
```diff
@@ -1404,45 +1404,121 @@ void IndexLowering::handleGroupedGridWelford(
...
+  Val* smem_address_ptr = nullptr;
+
+  if (minval->mbarrier()->isA<TensorView>()) {
+    smem_address_ptr =
+        lower_utils::u32IndexScalarSmemTv(minval->mbarrier()->as<TensorView>());
+  } else if (minval->mbarrier()->isA<kir::TensorIndex>()) {
+    smem_address_ptr = lower_utils::u32IndexScalarSmemTv(
+        minval->mbarrier()->as<kir::TensorIndex>());
+  } else {
+    NVF_ERROR(false, "Unexpected MBarrierInval value.");
```
Comment (jacobhinkle) [inline csrc/device_lower/pass/index.cpp:1434]:
```suggestion
    NVF_ERROR(false, "Unexpected MBarrierInvalidate barrier value: ", minval->mbarrier()->toString());
```

File: csrc/device_lower/pass/index.cpp [hunk 5]
```diff
@@ -1404,45 +1404,121 @@ void IndexLowering::handleGroupedGridWelford(
...
-      minit->threadCount());
+  Val* smem_address_ptr = nullptr;
+
+  if (minit->mbarrier()->isA<TensorView>()) {
+    smem_address_ptr =
+        lower_utils::u32IndexScalarSmemTv(minit->mbarrier()->as<TensorView>());
+  } else if (minit->mbarrier()->isA<kir::TensorIndex>()) {
+    smem_address_ptr = lower_utils::u32IndexScalarSmemTv(
+        minit->mbarrier()->as<kir::TensorIndex>());
+  } else {
```
[the remark this answers is not in the document]
Comment (rdspring1) [reply csrc/device_lower/pass/index.cpp:1415]: Perhaps. I was following a rule of 3. Two special cases is okay. Add dispatch if we have a third special case.

File: csrc/device_lower/utils.h (hunk 1 above)
Comment (rdspring1) [reply csrc/device_lower/utils.h:307]: A single `mbarrier` and its token are represented as a size-0 `int64_t` TensorView in shared memory, which can be called a scalar, smem TV.

File: csrc/device_lower/utils.h (hunk 3 above)
Comment (rdspring1) [reply csrc/device_lower/utils.h:309]: It feels unnecessary but this signature is derived from `NVF_API Val* u32IndexScalarSmemTv(TensorView* tv);`

File: csrc/device_lower/pass/index.cpp (hunk 2 above)
Comment (rdspring1) [reply csrc/device_lower/pass/index.cpp:1495]: The TMA tests without circular buffering will exercise the `else` branch. The code for `handleCpAsyncBulkLoad` was moved into the `else` branch in this PR.

File: csrc/device_lower/utils.h (hunk 3 above)
Comment (rdspring1) [reply csrc/device_lower/utils.h:309]: I'll try this change in a separate PR.

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Pull request merged**